### PR TITLE
[gpt-oss] Cache permute indices for faster MXFP4 MoE layer loading

### PR DESCRIPTION
**Pull request by frank-wei**
Summary:
On GB200, the MOE MXFP4 weight transpose takes quite a long time when the gpt-oss model is loaded.
Add the cache for weight transpose indices so that the expert weight transpose time can be reduced

**20b:**
Before: Model loading took 94sec
```
^[[1;36m(EngineCore_0 pid=3397977)^[[0;0m INFO 09-01 19:27:08 [default_loader.py:267] Loading weights took 2.83 seconds
^[[1;36m(EngineCore_0 pid=3397977)^[[0;0m INFO 09-01 19:28:41 [gpu_model_runner.py:1977] Model loading took 14.1643 GiB and 94.110470 seconds
```
After: Model loading took  5.9sec
```
^[[1;36m(EngineCore_0 pid=3005216)^[[0;0m INFO 09-02 16:54:43 [default_loader.py:267] Loading weights took 2.54 seconds
^[[1;36m(EngineCore_0 pid=3005216)^[[0;0m INFO 09-02 16:54:47 [gpu_model_runner.py:1977] Model loading took 14.1693 GiB and 5.918206 seconds
```

**120b:**
**Loading time verification:**
**Before, P1928776629**
E2E predictor warm up takes: 17:28:53 ~  17:39:59 = 11min 6sec

Model loading takes  568.133048 seconds
```
(EngineCore_0 pid=344869) INFO 09-02 17:29:45 [default_loader.py:267] Loading weights took 8.25 seconds
(EngineCore_0 pid=344869) INFO 09-02 17:39:05 [gpu_model_runner.py:1977] Model loading took 68.7019 GiB and 568.133048 seconds
```

**After, P1928762318**
E2E predictor warm up takes: 17:26:12 ~ 17:28:15 = 2min 3sec

Model loading takes 15.083996 seconds
```
(EngineCore_0 pid=156514) INFO 09-02 17:27:05 [default_loader.py:267] Loading weights took 9.18 seconds
(EngineCore_0 pid=156514) INFO 09-02 17:27:12 [gpu_model_runner.py:1977] Model loading took 68.7093 GiB and 15.083996 seconds
```
**Accuracy verification:**
```
aime25 medium: P1928806083
[{'eval_name': 'aime25', 'model_name': 'gpt-oss-120b-medium_temp1.0_20250902_175112', 'metric': 0.7875}]

aime25 high:P1928898566
[{'eval_name': 'aime25', 'model_name': 'gpt-oss-120b-high_temp1.0_20250902_180141', 'metric': 0.9}]
```

Test Plan:
Compared the transposed weights and they are matched between before and after. [[link](https://gist.github.com/frank-wei/0dd5c4d90f1bec3c3ce016230e5e66bc)]
python test_eq.py
```
import torch

[g1w, g1s, g1b] = torch.load("/tmp/gemm1_wei.pt")
[g1w2, g1s2, g1b2] = torch.load("/tmp/gemm1_wei2.pt")

for i in range(len(g1w)):
    print(i)
    print(torch.equal(g1w[i], g1w2[i]))
    print(torch.equal(g1s[i], g1s2[i]))
    print(torch.equal(g1b[i], g1b2[i]))

[g2w, g2s, g2b] = torch.load("/tmp/gemm2_wei.pt")
[g2w2, g2s2, g2b2] = torch.load("/tmp/gemm2_wei2.pt")

for i in range(len(g2w)):
    print(i)
    print(torch.equal(g2w[i], g2w2[i]))
    print(torch.equal(g2s[i], g2s2[i]))
    print(torch.equal(g2b[i], g2b2[i]))
```

Rollback Plan:

Differential Revision: D81544286

### CODE REVIEW

**Comment by 22quinn** [comment]
Look good! Let's add a unit test?

**Comment by yeqcharlotte** [comment]
thanks for the change! this is huge! 

could you update your PR title and replace internal pastebin with gist? :) 

cc: @houseroad @mgoin @LucasWilkinson

**Comment by frank-wei** [comment]
Thanks @yeqcharlotte and @22quinn for the review. I have updated this PR as suggested.

**Comment by jwfromm** [comment]
Great fix! Just curious, do we know why this issue is so much more noticeable on GB200 than other GPUs? It seems like this improvement is backend agnostic.

**Comment by frank-wei** [comment]
> Great fix! Just curious, do we know why this issue is so much more noticeable on GB200 than other GPUs? It seems like this improvement is backend agnostic.

@jwfromm , the issue is raised during enabling the mxfp4 on MOE weight. AFAIK, only Blackwell supports this format.

**Review by 22quinn — APPROVED** [review]
